Fix a bug where first column is clipped if it has fixed or static width

tnajdek opened the pull request:
This PR addresses issues #3582 and #3595.

The solution implemented here is functional, but there may be room for a more elegant approach. Given number of reports we had for the first column bug, we could opt to merge this as a quick fix and refine it later.

Changes include:

* A new handler `onSetOrder` in the virtualized table.
* Virtualized table now invokes onResize after the order of columns changes.
* The Items table uses `onSetOrder` to adjust the width of fixed/static width columns when they become the first columns.
* Similarly, when a column is toggled between visible and hidden states, and it becomes the first fixed/static column, its width needs adjustment.
* The exception for attachment column to use a different box-sizing has been removed. The fixed width has been adjusted to include column padding.


https://github.com/zotero/zotero/assets/214628/8bbbd308-939d-4218-92f6-d18038314c35

dstillman commented:
This makes the Notes column much wider than it seems like it needs to be:

![Screenshot 2024-01-27 at 3 29 47 AM](https://github.com/zotero/zotero/assets/783082/fba87ac7-e2ab-4abc-8e2a-9a8b74f9f435)

In the usual position at the end, it just adds a whole lot of wasted space on the right side of the table:

![Screenshot 2024-01-27 at 3 31 03 AM](https://github.com/zotero/zotero/assets/783082/8ad71c4f-6e5c-4280-ac8d-c4a617da5179)

dstillman commented:
And yes, while we can get this in first, we definitely will need to do something different on the left — at least different spacing/positioning of the attachment icon. Also looks like the attachment icon isn't properly vertically centered?

@yexingsha

tnajdek commented:
> This makes the Notes column much wider than it seems like it needs to be:

Old notes column had min width of 14px. New icon is 16px and column padding is 8px, thus making it 32px min width. We can add an exception for this column to allow no padding, bringing it down to 16px min width.

>  Also looks like the attachment icon isn't properly vertically centered?

I think attachment icon is off by 2px when it's first column, I'll add an exception.

dstillman commented:
Yeah, removing padding sounds good.

tnajdek commented:
I think this is much better now:

* I've set `minWidth` of notes column to 26px. It feels like a sweet spot and very close to what we had before (we had 14px coded but that was actually 24px because `onResize` added 10px of column padding)
* I've fixed the issue with attachment icon being off-center when it was configured as a first column
* I've covered few more cases where column widths needed to be adjusted. To keep code cleaner I've moved the logic to `Columns` class inside virtualized table.

dstillman commented:
Hmm, I'm still seeing the icon not vertically centered. To be clear, this is what I'm talking about:

![attachment-icon-vertical-centering](https://github.com/zotero/zotero/assets/783082/3b3a381e-3172-4741-b042-917e0bc1bfe8)

First row is the build and second row is fixed.

First row: 3px/1px offset for the journal article icon and 0px/2px for the P

Second row: 2px/2px for the journal article icon and 1px/1px for the P

dstillman commented:
Note column min-width looks good.

tnajdek commented:
> Hmm, I'm still seeing the icon not vertically centered. To be clear, this is what I'm talking about:

Thanks for clarifying! For some reason I thought this was about row column vs header column alignment.

I've now centered it, like this:

<img width="1387" alt="Screenshot 2024-01-28 at 10 21 48" src="https://github.com/zotero/zotero/assets/214628/e4e44208-c1f7-4bcf-b50a-c18abc6ce39f">

dstillman commented:
That is much better! But also not what I am referring to.

I think you just missed the "vertically" when I said "vertically centered" above? Not horizontal. My previous screenshot shows me moving the attachment icon up 1px in the second row — that's all I'm talking about. The offsets I gave are the offsets from the top/bottom of the item-type icon and the "P". Sorry for not being clearer.

(But I was going to mention the horizontal off-center-ness too, so this is good! Still maybe a little weird, and @yexingsha might have other suggestions, but definitely an improvement.)

tnajdek commented:
Sorry, I rushed into the work and I missed your point.

I've now centered the icon vertically, at this scale inspector isn't very helpful (guides are too thick) so here is a zoomed screenshot with borders added in post-edit (both red lines are 4px long, this is a screenshot from a 2x screen so that's equivalent to 2px)

![vertical](https://github.com/zotero/zotero/assets/214628/ef79d857-3264-4e8a-b100-c6f500ea42f5)

tnajdek commented on chrome/content/zotero/components/virtualized-table.jsx:
```diff
@@ -848,8 +849,8 @@ class VirtualizedTable extends React.Component {
-		const aSpacingOffset = (aColumn.minWidth ? aColumn.minWidth : COLUMN_MIN_WIDTH) + COLUMN_PADDING;
-		const bSpacingOffset = (bColumn.minWidth ? bColumn.minWidth : COLUMN_MIN_WIDTH) + COLUMN_PADDING;
+		const aSpacingOffset = (aColumn.minWidth ? aColumn.minWidth : COLUMN_MIN_WIDTH) + (aColumn.noPadding ? 0 : COLUMN_PADDING);
+		const bSpacingOffset = (bColumn.minWidth ? bColumn.minWidth : COLUMN_MIN_WIDTH) + (bColumn.noPadding ? 0 : COLUMN_PADDING);
```
I'd argue it's explicit, not finicky. That being said, having icon columns imply no padding here makes sense, no objections there, ideally with few words of comment in jsdoc to make everything clear.
